nevergrad/functions/rocket/rocket.py

```diff
@@ -0,0 +1,286 @@
+    thrust_time_list = np.asarray([thrust[i+1][1] - thrust[i][1] for i in range(0, len(thrust)-1)])
+    total_thrust = np.sum(np.multiply(thrust_list, thrust_time_list))
+    
+    # We moodify the thrust while preserving the sum (this is an adaptation to Nevergrad).
+    # 1: we modify.
+    thrust_list = np.multiply(thrust_list, np.exp(thrust_bias))
+    # 2: we normalize.
+    thrust_list = thrust_list * total_thrust / np.sum(np.multiply(thrust_list, thrust_time_list))
+    
+    for i in range(len(thrust)-1):
+        thrust[i][0] = thrust_list[i]
+    # total_mass vs time curve
+    # this is used to represent the mass loss while the rocket burns fuel
+    mass_time = []
+    #total_thrust = 0
+    #for row in thrust:
+    #    total_thrust += row[0]
+    
+    mass_loss = eng_mass_initial - eng_mass_final
+    mass_reman = eng_mass_initial
+    for row in thrust:
+        # Equation below weird to me: this is not normalized by time ? the mass which is lost should be proportional
+        # to thrust x delta-time, right ?
+        #percentage = row[0] / total_thrust   # percentage of total thrust to find percentage of mass lost
+        percentage = row[0] / np.sum(thrust_list)   # percentage of total thrust to find percentage of mass lost
+        assert percentage >= 0.
+        assert percentage <= 1.
+        mass_loss = mass_reman * percentage
+        mass_reman -= mass_loss
+        total_mass = roc_mass + mass_reman
+        mass_time.append([total_mass, row[1]])
+    mass_list = [mass_time[i][0] for i in range(0, len(thrust))]
+    
+    # Lists used to store data and later import data to excel
+    Ex_list, Ey_list, Ez_list = [Ex], [Ey], [Ez]
+    Evx_list, Evy_list, Evz_list = [Evx], [Evy], [Evz]
+    time_list = [0]
+    r_list = [(Ex ** 2 + Ey ** 2 + Ez ** 2) ** 0.5]
+    
+    # Initializing variables
+    time = 0.  # time in seconds
+    
+    
+    # while thrust is greater than zero
+    # this is while the rocket engine is firing
+    for i in range(len(thrust) - 2):
+        r = (Ex ** 2 + Ey ** 2 + Ez ** 2) ** 0.5
+        dt = thrust[i][1]
+        Efx, Efy, Efz = net_force(Ex, Ey, Ez, Evx, Evy, Evz, mass_list[i])
+        Ex += (Evx * dt)
+        Ey += (Evy * dt)
+        Ez += (Evz * dt)
+        dt = thrust[i + 1][1] - thrust[i][1]
+        Evz += ((((thrust[i][0] * math.cos(theta)) + Efz) * dt) / mass_list[i])
+        Evx += ((((thrust[i][0] * math.sin(theta)*math.cos(phi)) + Efx) * dt) / mass_list[i])
+        Evy += ((((thrust[i][0] * math.sin(theta)*math.sin(phi)) + Efy) * dt) / mass_list[i])
+        time += dt
+        update(Ex, Ey, Ez, Evx, Evy, Evz, time, r)
+    
+    # After thrust
+    # This is when the engine is out of fuel and there is no longer a thrust force
+    time_step = .05  # time time_step in seconds
+    dt = time_step
+    while r > r_initial:
+        r = (Ex ** 2 + Ey ** 2 + Ez ** 2) ** 0.5
+        Efx, Efy, Efz = net_force(Ex, Ey, Ez, Evx, Evy, Evz, final_roc_mass)
```

Comment:
this all looks like it should be numpy arrays to:
- require less lines of code
- be much faster

Reply:
Replaced lists by numpy arrays.
Did not optimize anything for now.